{
  "cells": [
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "title: \"Procesamiento bioinformático: demultiplex - llamado de SNPs\"\n",
        "format: html\n",
        "author: \"Sergio y Sofía\"\n",
        "self-contained: true\n",
        "keep-md: true\n",
        "---"
      ],
      "id": "742638b5"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Instalar miniconda en su usuario \n",
        "\n",
        "\n",
        "## Demultiplex\n",
        "\n",
        "### process_radtags\n",
        "\n",
        "STACKS tool process_radtags which:\n",
        "\n",
        "> ... checks that the barcode and the RAD cutsite are intact, and demultiplexes the data. If there are errors in the barcode or the RAD site within a certain allowance process_radtags can correct them. Second, it slides a window down the length of the read and checks the average quality score within the window. If the score drops below 90% probability of being correct (a raw phred score of 10), the read is discarded. This allows for some seqeuncing errors while elimating reads where the sequence is degrading as it is being sequenced.\n",
        "\n",
        "- [Manual](https://catchenlab.life.illinois.edu/stacks/comp/process_radtags.php)\n",
        "- script: `process_radtags.sh`\n",
        "\n",
        "Parameters:\n",
        "- `-p` we indicate the directory with the sequence data\n",
        "- `-i` the type of sequence file\n",
        "- `-o` the output directory\n",
        "- `-y` the output format\n",
        "- `-b` the barcode file to use for demultiplexing\n",
        "- `-e` the restriction enzyme so it knows what cut site sequence to expect after the barcode\n",
        "- `--adapter_1` to indicate the adapter sequence to identify and remove contaminants\n",
        "- `--adapter_mm` the number of mismatches allowed between the read and real adapter sequence\n",
        "- `-r` to \"rescue\" (keep) a sequence if there is one error in the barcode\n",
        "- `-c` to remove reads with uncalled bases (N)\n",
        "- `-q` to discard reads with low quality scores which are defined by a sliding window if the average score is less than `-s` = 10.\n",
        "\n",
        "\n",
        "```{bash}\n",
        "\n",
        "input_dir=../data/raw_data/batch_01846\n",
        "barcodes_file=../data/barcodes/barcodes_01846.txt\n",
        "\n",
        "#mkdir ../data/demultiplexed_raw2/batch_01846\n",
        "\n",
        "process_radtags -t 18 \\ # número de núcleos\n",
        "                -p $input_dir \\\n",
        "                -i gzfastq \\\n",
        "                -o ../data/demultiplexed_raw2/batch_01846 \\\n",
        "                -b $barcodes_file \\\n",
        "                --renz_1 pstI \\ # enzima de restricción\n",
        "                -r -c \\\n",
        "                --barcode-dist-1 2 \\ # numero de discrepancias en el barcode\n",
        "                -D \\\n",
        "                &>out_demultiplex_01846.log # una forma de guardar lo que se imprime en pantalla \n",
        "```\n",
        "\n",
        "\n",
        "\n",
        "## Revisión de calidad\n",
        "\n",
        "\n",
        "\n",
        "### FastQ\n",
        "\n",
        "\n",
        "```{bash}\n",
        "files=$(ls -d ../data/raw_data/morton_data/*.gz)\n",
        "\n",
        "for file in $files\n",
        "do\n",
        "fastqc $file -o ../outputs/fastqc/morton_data/\n",
        "done\t\n",
        "\n",
        "```\n",
        "\n",
        "\n",
        "### FastQC\n",
        "\n",
        "\n",
        "```{bash}\n",
        "files=$(ls -d ../data/raw_data/morton_data/*.gz)\n",
        "\n",
        "fastq_screen --threads 10 --aligner bowtie2 --conf fastq_screen.conf $files --outdir ../outputs/fastq_screen/morton_data/\n",
        "\n",
        "```\n",
        "\n",
        "\n",
        "### MultiQC - Resumen general\n",
        "\n",
        "```{.bash}\n",
        "\n",
        "## Directories\n",
        "\n",
        "#fastqc=\"../outputs/fastqc/batch_01846/\"\n",
        "fastqc=\"../outputs/fastqc/morton_data/\"\n",
        "\n",
        "#fastq_screen=\"../outputs/fastq_screen/batch_01846/\"\n",
        "fastq_screen=\"../outputs/fastq_screen/morton_data/\"\n",
        "\n",
        "## Command\n",
        "\n",
        "multiqc --force --interactive $fastqc $fastq_screen --outdir ../outputs/multiqc/morton_data/ \n",
        "\n",
        "```\n",
        "\n",
        "\n",
        "## Mapeo a un genoma de referencia\n",
        "\n",
        "## Aligment\n",
        "\n",
        "After initial quality and contamination assesment. Next step is to performe an aligment to a reference genome (in case there is one available,\n",
        "otherwise a _de Novo_ aligment and variant callig pipeline has to be used). Further trimming and quality chechs can be evaluated by looking at their\n",
        "effect on the mapping statistics.\n",
        "\n",
        "First the reference genome has to be downloaded. See [genome_ncbi_download.md](./genome_ncbi_download.md) for short instructions. The reference\n",
        "needs to be in a fasta format (.fna is fasta format when sequences are downloaded form NCBI).\n",
        "\n",
        "-----\n",
        "\n",
        "#### Instructions for simple genome download from NCBI\n",
        "\n",
        "In the NCBI website reference genomes are refered as assemblies (representation of genomes). Assemblies is a database in the site where genomes can \n",
        "be looked for. \n",
        "\n",
        "There are 2 options to download ncbi genomes (both use a FTP connection) through rsync commands or wget/curl commands. There is a short \n",
        "[tutorial](https://www.youtube.com/watch?v=-X0H024ST8k&ab_channel=TheNationalLibraryofMedicine) and documentation in the [FTP \n",
        "QA](https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/)\n",
        "\n",
        "##### Small number of genomes to be downloaded - rsync\n",
        "\n",
        "The FTP path can be obtained by first looking up the species of interes in the NCBI Assembly database, click on the record and on the right side \n",
        "panel there is an option for FTP directories. Once you get the FTP path it can be inserted in the rsync command. \n",
        "\n",
        "Replace the \"ftp:\" at the beginning of the FTP path with \"rsync:\". The directory and its contents could be downloaded using the following rsync command:\n",
        "\n",
        "`rsync --copy-links --recursive --times --verbose rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/633/185/GCF_001633185.2_ValleyOak3.2/`\n",
        "\n",
        "-----\n",
        "\n",
        "Then the reference genome has to be indexed. This basically means that we can create a system to do queries more efficient. To uderstand more deeply\n",
        "[Lectures](https://www.youtube.com/watch?v=5G2Db41pSHE&list=PL2mpR0RYFQsADmYpW2YWBrXJZ_6EL_3nu&index=1&ab_channel=BenLangmead).\n",
        "\n",
        "```{.bash}\n",
        "# index reference genomes to use fastq_screen\n",
        "\n",
        "cd ../data/genomes/\n",
        "\n",
        "paths=$(ls -d $PWD/*)\n",
        "\n",
        "for i in $paths\n",
        "do\n",
        "file=$(ls -d $i/*.fna.gz)\n",
        "name=$(ls $i/*.fna.gz | cut -d . -f 1-2)\n",
        "cd $i\n",
        "bowtie2-build $file $name\n",
        "done\n",
        "\n",
        "```\n",
        "After that we can use one of the bwa algorithms to map short sequences into our reference.\n",
        "\n",
        "```{.bash}\n",
        "\n",
        "## Align to reference genome (Q.lobata) using BWA-MEM\n",
        "\n",
        "######### read group file suffix\n",
        "\n",
        "#file_suffix=\"01846\"\n",
        "file_suffix=\"morton_data2\"\n",
        "#file_suffix=\"all.90\"\n",
        "\n",
        "######### input data directory\n",
        "\n",
        "#folder_data=\"../data/demultiplexed_raw2/all.90bp\"\n",
        "folder_data=\"../data/raw_data/morton_data/set2/\"\n",
        "\n",
        "######### output directories\n",
        "\n",
        "out_folder=\"morton_data2\"\n",
        "\n",
        "mkdir -p ../data/alignment/sam/$out_folder\n",
        "mkdir -p ../data/alignment/bam/$out_folder\n",
        "\n",
        "while read sample readgroup\n",
        "do\n",
        "\n",
        "echo \"bwa mem -M -t 18 -R ${readgroup} ../data/genomes/qlobata/GCA_001633185.5_ValleyOak3.2_genomic.fna ${folder_data}/${sample}.fastq.gz > ../data/alignment/sam/${out_folder}/${sample}.sam\" | bash\n",
        "\n",
        "picard SortSam I=../data/alignment/sam/${out_folder}/${sample}.sam O=../data/alignment/bam/${out_folder}/${sample}.bam SORT_ORDER=coordinate\n",
        "\n",
        "samtools index ../data/alignment/bam/${out_folder}/${sample}.bam\n",
        "\n",
        "done < ../data/readgroups/sampl_rgp_${file_suffix}\n",
        "\n",
        "```\n",
        "\n",
        "## Llamado de SNPs\n"
      ],
      "id": "b37b0674"
    }
  ],
  "metadata": {
    "kernelspec": {
      "name": "python3",
      "language": "python",
      "display_name": "Python 3 (ipykernel)"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}